conjur/user/user_logic.py

```diff
@@ -0,0 +1,100 @@
+# -*- coding: utf-8 -*-
+
+"""
+UserLogic module
+
+This module is the business logic for handling all user-related activity
+"""
+
+# Builtins
+import logging
+
+# Internals
+from conjur.credentials_data import CredentialsData
+
+class UserLogic:
+    """
+    UserLogic
+
+    This class holds the business logic for handling
+    user activity
+    """
+    def __init__(self, conjurrc_data, credentials_from_store, client, resource):
+        self.conjurrc_data=conjurrc_data
+        self.credentials_from_store=credentials_from_store
+        self.client=client
+        self.resource=resource
+
+    # pylint: disable=logging-fstring-interpolation
+    def rotate_api_key(self, resource_to_update):
+        """
+        Method to trigger two types of API key actions.
+        1. The user can either rotate their own API key or
+        2. Rotate another users API key
+        """
+        loaded_credentials = self.extract_credentials_from_credential_store()
+        user_from_credential_store = loaded_credentials['login_id']
+        if resource_to_update == user_from_credential_store:
+            raise Exception("Error: To rotate the API key of the currently logged-in user "\
+                            " use this command without any flags or options")
+
+        if resource_to_update is not None:
+            user_from_credential_store = resource_to_update
+            logging.debug(f"Rotating API key for '{user_from_credential_store}'")
+            new_api_key = self.rotate_anothers_api_key(user_from_credential_store)
+            logging.debug(f"API key for '{user_from_credential_store}' successfully rotated")
+        else:
+            # if the user does not provide a user to rotate their API key,
+            # their own API key will be rotated
+            logging.debug(f"Rotating API key for '{user_from_credential_store}'")
+            new_api_key = self.rotate_personal_api_key(user_from_credential_store,
+                                                       loaded_credentials['api_key'])
+
+            # Update the new rotated API for the logged-in user
+            logging.debug("Updating credential store with new API key " \
+                          f"for '{user_from_credential_store}'")
+            self.update_api_key_in_credential_store(user_from_credential_store,
```

Comment:
incase an error occurs between setting the new api_key and update it in the store (again good encapsulation). we are in an unstable situation where the api key at the server is different from the store. maybe we should alert the user? or try to fetch the key from the server 
```suggestion
    try:
            new_api_key = self.rotate_personal_api_key(user_from_credential_store,
                                                       loaded_credentials['api_key'])

            # Update the new rotated API for the logged-in user
            logging.debug("Updating credential store with new API key " \
                          f"for '{user_from_credential_store}'")
            self.update_api_key_in_credential_store(user_from_credential_store,
     except:
         reRaise exception
    finally:
        #get api key from server and store it on the store
```

Reply:
Correct it would lead to an unstable situation and needs to be addressed! One thing though, I am not understanding how the above suggestion is overcoming this instability because in the end we are just making the same request twice
1. Rotate API key
2. Attempt to store
3. Fail
4. Rotate API key (again)
5. Attempt to store (again)
There is no way to get the API key again, only rotate.

How do you feel about the following flow
1. Rotate API key
2. Attempt to store
3. Fail
4. Return saying that the API key was rotated and it is AX12 but something went wrong and is in an unstable state.

Reply:
sounds good. 
one thing that i don't understand. if the rotate failed than we can't perform login and get the api key.
I think that the error message to the user should be something like "rotate failed please login again to get the currect api-key."
but if we expect the user to login afterwards than we can to it ourselves in the `except` or `finally` stage as part of the function error handling .

Reply:
The general statement is `Failed to execute command. Reason: 401 Client Error: Unauthorized for url: https://localhost/authn/cucumber/adfmin/authenticate` `Failed to execute command` is general enough so I dont think we need to say rotate failed because we failed on authentication